### PR TITLE
Fix #153/#154 : MultipleChoice/ItemSelection Interaction View

### LINKED ISSUE #153: UI Structure: MultipleChoiceInteractionView [Blocked: #150]
This is tracking implementing the initial UI structure of the multiple choice interaction being fully implemented in #31.

### PR DESCRIPTION
**Pull request by veena14cs**
##  Explanation

This PR is duplicate of [#184](https://github.com/oppia/oppia-android/pull/184). It contains UI structure for Multiple choice and item selection interactions, and fixes #153 and #154.

Reference Design Doc: https://docs.google.com/document/d/1HFwDFcqenAXepbYr5rfawGHvkeBZpYvNMs228AWLn24/edit?usp=sharing

### CODE REVIEW

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import android.content.Context
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import org.oppia.app.fragment.InjectableFragment
+import org.oppia.app.player.content_interaction.ContentInteractionFragmentPresenter
+import javax.inject.Inject
+
+/** Fragment displays single/multiplechoice input interaction. */
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt]: Do not you `single` or `singlechoice` or `singlechoiceinput` anywhere, in our discussion that was just for communication.
If you are using `SingleChoiceInput` -> that by default developer will understand that in this radio buttons should be visible and leaner can select only one correct answer. This then drives the assumption that `MultipleChoiceInput` will show checkboxes and learner can select multiple answers.

This is incorrect in terms of definition as well as oppia-web codebase.

Use `ItemSelectionInput` or `ItemSelectionViewInteraction` whichever fits the name, this means that learner can select multiple answers at a time in backend code.
Use `MultipleChoiceInput` or `MultipleChoiceViewInteraction`, this means that learner can select one answer among multiple options, as per backend code.

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+class ContentInteractionFragment : InjectableFragment() {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt]: Nit: Remove this empty line.

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+import android.widget.Toast
+import kotlinx.android.synthetic.main.content_interaction_fragment.view.*
+import org.oppia.util.data.UrlImageParser
+import android.widget.CompoundButton
+
+/** Presenter for [ContentInteractionFragment]. */
+class ContentInteractionFragmentPresenter @Inject constructor(
+  @ApplicationContext private val context: Context,
+  private val fragment: Fragment
+//  private val interactionInstanceId: String, val gaeCustomizationArgs: Map<String, GaeCustomizationArgs>
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Remove this comment and unused variable

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+  private lateinit var binding: ContentInteractionFragmentBinding
+
+  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
+
+    binding = ContentInteractionFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
+
+    showInputInteractions()
+
+    return binding.root
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Remove extra empty spaces.

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  private fun showInputInteractions() {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Remove extra empty space

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    //Todo(Veena): Remove static initialization and use values from constructor for gaeCustomizationArgsMap and interactionInstanceId
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Space just before TODO.

This should be 
`// Todo(Veena): Remove static ...`

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
+
+    binding = ContentInteractionFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
+
+    showInputInteractions()
+
+    return binding.root
+  }
+
+  private fun showInputInteractions() {
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: It would be good, if your method accepts whatever parameters it needs and then create the UI. So that, when we work on final implementation we can directly use your method.

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      var items = gaeCustomArgsInString.split(",").toTypedArray()
+      addCheckbox(items)
+    } else {
+      //Do no show any view
+    }
+  }
+
+  fun addCheckbox(optionsArray: Array<String>) {
+    for (row in 0..0) {
+      val ll = LinearLayout(context)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Rather than using `ll` using proper variable names like `checkboxLinearLayout` or `rootLinearLayout` or simply `linearLayout`

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+      ll.orientation = LinearLayout.VERTICAL
+
+      for (i in 0..optionsArray.size - 1) {
+        val cb = CheckBox(context)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Change variable name as per above comment.

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        
+        ll.addView(cb)
+      }
+      binding.root.radioGroup.addView(ll)
+    }
+  }
+
+  fun addRadioButtons(optionsArray: Array<String>) {
+    for (row in 0..0) {
+      val rg = RadioGroup(context)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Nit: Variable name

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+      rg.orientation = LinearLayout.VERTICAL
+
+      for (i in 0..optionsArray.size - 1) {
+        val rdbtn = RadioButton(context)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Nit: Variable name

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        rdbtn.id = View.generateViewId()
+        rdbtn.text = convertHtmlToString(optionsArray[i], rdbtn).toString()// + rdbtn.id
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Remove comment if not needed

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+          override fun onCheckedChanged(group: RadioGroup, checkedId: Int) {
+            for (i in 0 until group.childCount) {
+              val btn = group.getChildAt(i) as RadioButton
+              if (btn.id == checkedId) {
+                val text = btn.text
+                Toast.makeText(context, "" + text, Toast.LENGTH_LONG).show()
+                return
+              }
+            }
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Remove this extra space

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+          }
+        })
+        rg.addView(rdbtn)
+      }
+      binding.root.radioGroup.addView(rg)
+    }
+  }
+
+  fun convertHtmlToString(rawResponse: Any?, rdbtn: View): Spanned {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Remove this extra space

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+    var htmlContent = rawResponse as String;
+    var result: Spanned
+    var CUSTOM_TAG = "oppia-noninteractive-image"
+    var HTML_TAG = "img"
+    var CUSTOM_ATTRIBUTE = "filepath-with-value"
+    var HTML_ATTRIBUTE = "src"
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: I think all these variables can be converted to val. But better you define all these as `const val` before above the entire-class-code

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+    if (htmlContent!!.contains(CUSTOM_TAG)) {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: Remove this extra line.

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+import android.graphics.drawable.LevelListDrawable
+import android.text.Html
+import android.widget.TextView
+import com.bumptech.glide.Glide
+import com.bumptech.glide.request.target.SimpleTarget
+import com.bumptech.glide.request.transition.Transition
+import java.net.URL
+
+/** UrlImage Parser for android TextView to load Html Image tag */
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: Remove this line. Also add `.` in the comment

`/** UrlImage Parser for android TextView to load Html Image tag. */`

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+    return urlDrawable
+  }
+
+  inner class BitmapTarget(private val urlDrawable: UrlDrawable) : SimpleTarget<Bitmap>() {
+    internal var drawable: Drawable? = null
+
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+
+      drawable = BitmapDrawable(context.resources, resource)
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: Nit: Remove extra space

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    targets?.add(target)
+    load.into(target)
+    return urlDrawable
+  }
+
+  inner class BitmapTarget(private val urlDrawable: UrlDrawable) : SimpleTarget<Bitmap>() {
+    internal var drawable: Drawable? = null
+
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: Nit: Remove extra space

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+      drawable = BitmapDrawable(context.resources, resource)
+
+      tvContents.post {
+        val w = tvContents.width
+        val hh = (drawable as BitmapDrawable).intrinsicHeight
+        val ww = (drawable as BitmapDrawable).intrinsicWidth
+        val newHeight = hh * w / ww
+        val rect = Rect(0, 0, w, newHeight)
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: Name variables correctly

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+        (drawable as BitmapDrawable).bounds = rect
+        urlDrawable.bounds = rect
+        urlDrawable.drawable = drawable
+        tvContents.text = tvContents.text
+        tvContents.invalidate()
+      }
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: Nit: Remove extra space

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  inner class UrlDrawable : BitmapDrawable() {
+    var drawable: Drawable? = null
+    override fun draw(canvas: Canvas) {
+      if (drawable != null)
+        drawable!!.draw(canvas)
+    }
+  }
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: Nit: Remove extra space

**Comment by rt4914** [comment]
@veena14cs 

Your code for checkbox and radio button works correct. Great job.

I have left some comments which you can resolve.

Also, you are using `ContentInteractionFragment` just for showing this demo and it actually serves no other purpose. And we know that we have to use these two views in the `StateFragment`, so I suggest you remove `ContentInteractionFragment` and shift your implementation in `StateFragment`.

Also, create a customview which handles your two views, checkout @nikitamarysolomanpvt PR on FractionInput, she has created a separate `CustomView` rather than handling everything in parent fragment.

Write test-cases for this code.

**Comment by rt4914** [comment]
@veena14cs  your branch name contains '()' brackets, do not add brackets to a branch name, it is wrong convention. Also, if the branch name contains `()` then switching branches from terminal is not possible directly.

**Review by rt4914 — COMMENTED** [review]
PTAL

File: app/src/main/res/layout/content_interaction_fragment.xml [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+    <LinearLayout
+      android:id="@+id/rootContainer"
+      android:layout_width="match_parent"
+      android:layout_height="match_parent"
+      android:orientation="vertical">
+      <RadioGroup
+        android:id="@+id/radioGroup"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+      android:padding="15dp">
```
Comment (rt4914) [inline app/src/main/res/layout/content_interaction_fragment.xml]: Indentation issue.

**Comment by veena14cs** [comment]
Moved Interaction implementation to statefragment.

**Comment by veena14cs** [comment]
used the terms MultipleChoiceViewInteraction and ItemSelectionViewInteraction
for interactions.
But these terms are really creating confusion though it is according to
webcode base, in my opinion we have to change these terms according to
their definations

**Comment by veena14cs** [comment]
corrected indentation

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+import android.widget.TextView
+import android.widget.Toast
+import androidx.core.content.ContextCompat
+import org.oppia.app.R
+import org.oppia.util.data.UrlImageParser
+import android.content.res.ColorStateList
+import android.os.Build
+import android.view.Gravity
+
+// TODO(#190): Move this to a custom.inputinteractionview.
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: If here `custom.inputinteractionview` means folders than I guess you can do that right now and it won't create any issue with @nikitamarysolomanpvt 's PR#190

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+  private var mContext: Context
+  private var optionContents: String? = null
+
+  constructor(context: Context, optionContents: String) : super(context) {
+    this.mContext = context
+    this.optionContents = optionContents
+    initViews()
+  }
+
+  //update default attributes of MultipleSelectionInputInteractionView here
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: Please correct this comment format.
`// Update default attributes of MultipleSelectionInputInteractionView here.`

Nit:
There should be space after `//` and `.` at the end of sentence

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+  fun initViews() {
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: Nit: There are two types of constructors in kotlin, primary and secondary.

Maybe you need those constructors here.

Referenced: https://blog.mindorks.com/primary-and-secondary-constructors-in-kotlin

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+
+    setOnCheckedChangeListener(CompoundButton.OnCheckedChangeListener { buttonView, isChecked ->
+      val msg = "You have " + (if (isChecked) "checked" else "unchecked") + " this Check it Checkbox."
+      Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
+    })
+  }
+
+  var colorStateList = ColorStateList(
+    arrayOf(
+      intArrayOf(android.R.attr.state_enabled) //enabled
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: Remove comment. Your `state_enabled` is self-explanatory.

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+    ),
+    intArrayOf(ContextCompat.getColor(context, R.color.oppiaDarkBlue))
+  )
+
+  fun convertHtmlToString(rawResponse: Any?, rdbtn: View): Spanned {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: Remove this extra space.

we don't add extra space on the first and last line of method because indentation automatically makes everything clear.

**Review by rt4914 — COMMENTED** [review]
PTAL

**Comment by rt4914** [comment]
https://github.com/oppia/oppia-android/pull/188#issuecomment-535888499

@veena14cs 

i understand that this naming convention is very confusing but it better if we get used to it so that we can communicate easily with @seanlip and @BenHenning , because they are used to these terms whereas we are completely new to this project and it is easier for us to follow this and difficult for them to change this naming now.

File: app/src/main/AndroidManifest.xml [hunk 29]
```diff
@@ -11,8 +11,8 @@
     android:roundIcon="@mipmap/ic_launcher_round"
     android:supportsRtl="true"
     android:theme="@style/OppiaTheme">
-    <activity android:name=".player.exploration.ExplorationActivity"/>
-    <activity android:name="org.oppia.app.home.HomeActivity">
+    <activity android:name=".player.exploration.ExplorationActivity">
+    <!--<activity android:name="org.oppia.app.home.HomeActivity">-->
```
Comment (rt4914) [inline app/src/main/AndroidManifest.xml]: Revert these changes, we don't need to make sure that the PR code uses your implementation.
Rather than that we want that your PR should be merged directly and whenever the develop runs the develop branch, homefragment should appear.

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt [hunk 30]
```diff
@@ -14,12 +15,12 @@ class ExplorationActivityPresenter @Inject constructor(private val activity: App
...
-        ExplorationFragment()
+        StateFragment()
       ).commitNow()
     }
   }
 
-  private fun getExplorationFragment(): ExplorationFragment? {
-    return activity.supportFragmentManager.findFragmentById(R.id.exploration_fragment_placeholder) as ExplorationFragment?
+  private fun getExplorationFragment(): StateFragment? {
+    return activity.supportFragmentManager.findFragmentById(R.id.exploration_fragment_placeholder) as StateFragment?
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt]: Revert these changes, we don't need to make sure that the PR code uses your implementation.
Rather than that we want that your PR should be merged directly and whenever the develop runs the develop branch, homefragment should appear.

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt (hunk 26 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: Rather than calling `initViews()` inside `init` better implement the entire code of `initViews` inside `init` itself and remove the `initViews method`

File: app/src/main/java/org/oppia/app/player/state/CustomRadioButton.kt [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+// TODO(#163): Move this to a StateCardFragment Low-fi PR.
+/** Custom Checkbox for MultipleSelectionInputInteractionView. */
+class CustomRadioButton(context: Context?,private val optionContents: String) : RadioButton(context) {
+
+  init {
+    initViews()
+  }
+
+  // Update default attributes of ItemSelectionInputInteractionView here.
+  fun initViews() {
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/CustomRadioButton.kt]: Ditto as per earlier comment.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt [hunk 32]
```diff
@@ -31,4 +43,63 @@ class StateFragmentPresenter @Inject constructor(
   fun setAudioFragmentVisible(isVisible: Boolean) {
     getStateViewModel().setAudioFragmentVisible(isVisible)
   }
+
+  private fun showInputInteractions() {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: As mentioned earlier, we do not need extra black space after opening and closing bracket of any function.

File: app/src/main/res/layout/state_fragment.xml [hunk 33]
```diff
@@ -15,20 +15,36 @@
...
+      android:id="@+id/linearLayout"
       android:layout_width="0dp"
       android:layout_height="wrap_content"
       android:layout_marginStart="24dp"
       android:layout_marginEnd="24dp"
       android:visibility="@{viewModel.isAudioFragmentVisible? View.VISIBLE: View.GONE}"
       app:layout_constraintEnd_toEndOf="parent"
-      app:layout_constraintStart_toStartOf="parent"
-      app:layout_constraintTop_toTopOf="parent">
+      app:layout_constraintStart_toStartOf="parent" app:layout_constraintTop_toTopOf="parent">
```
Comment (rt4914) [inline app/src/main/res/layout/state_fragment.xml]: Correct this code indentation.

File: app/src/main/res/layout/state_fragment.xml [hunk 34]
```diff
@@ -15,20 +15,36 @@
...
     </LinearLayout>
+    <LinearLayout
+      android:id="@+id/interactionContainer"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_margin="10dp"
+      android:background="@color/white" android:orientation="vertical"
+      android:padding="10dp"
+      app:layout_constraintBottom_toTopOf="@+id/dummy_audio_button"
+      app:layout_constraintTop_toBottomOf="@+id/linearLayout" tools:layout_editor_absoluteX="8dp">
```
Comment (rt4914) [inline app/src/main/res/layout/state_fragment.xml]: Correct this code indentation

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt [hunk 35]
```diff
@@ -0,0 +1,101 @@
...
+//      onView(ViewMatchers.withHint("text"))
+//          .perform(clearText(), typeText("123"));
+//      pauseTestFor(500)
+//      onView(ViewMatchers.withHint("fraction 1/1"))
+//          .check(matches(isDisplayed()))
+//      onView(ViewMatchers.withHint("fraction 1/1"))
+//          .perform(click())
+//      onView(ViewMatchers.withHint("text"))
+//          .perform(clearText(), typeText("1/3"));
+//      pauseTestFor(500)
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt]: Why is this test commented? Also, this is not from your PR, this is from #190

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt [hunk 36]
```diff
@@ -0,0 +1,101 @@
...
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.oppia.app.R
+import org.oppia.app.home.HomeActivity
+import org.oppia.util.threading.BackgroundDispatcher
+import org.oppia.util.threading.BlockingDispatcher
+
+@RunWith(AndroidJUnit4::class)
+class StateFragmentPresenterTest {
+  //to test text,number,and fraction input interaction views
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt]: This comment is from PR #190

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 37]
```diff
@@ -0,0 +1,63 @@
...
+    targets?.add(target)
+    load.into(target)
+    return urlDrawable
+  }
+
+  inner class BitmapTarget(private val urlDrawable: UrlDrawable) : SimpleTarget<Bitmap>() {
+    internal var drawable: Drawable? = null
+
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: Remove this blank line.

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 38]
```diff
@@ -0,0 +1,63 @@
...
+      drawable = BitmapDrawable(context.resources, resource)
+      tvContents.post {
+        val w = tvContents.width
+        val hh = (drawable as BitmapDrawable).intrinsicHeight
+        val ww = (drawable as BitmapDrawable).intrinsicWidth
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: This issue was mentioned earlier too. Rename these variables to something which become self-explanatory.

**Review by rt4914 — COMMENTED** [review]
PTAL

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt (hunk 1 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt]: this file is been removed

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt (hunk 2 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragment.kt]: File is removed

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt (hunk 3 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: removed

File: app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt (hunk 5 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content_interaction/ContentInteractionFragmentPresenter.kt]: removed extra spaces

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt (hunk 18 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: removed spaces

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt (hunk 19 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: removed extra spaces

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt (hunk 20 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: variable names corrected

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt (hunk 22 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: removed spaces

File: utility/src/main/java/org/oppia/util/data/URlImageParser.kt (hunk 21 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/URlImageParser.kt]: removed blank spaces

File: app/src/main/res/layout/content_interaction_fragment.xml (hunk 23 above)
Comment (veena14cs) [reply app/src/main/res/layout/content_interaction_fragment.xml]: this file is been removed

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt (hunk 24 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: as per our conversations added TODO #163 StateCardFragment Low-fi PR.

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt (hunk 25 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: corrected

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt (hunk 26 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: added entire code to init()

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt (hunk 27 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: removed comment

File: app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt (hunk 28 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt]: ok removed blank space

File: app/src/main/AndroidManifest.xml (hunk 29 above)
Comment (veena14cs) [reply app/src/main/AndroidManifest.xml]: reverted

File: app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt (hunk 30 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt]: reverted

File: app/src/main/java/org/oppia/app/player/state/CustomRadioButton.kt (hunk 31 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/CustomRadioButton.kt]: corrected comment.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt (hunk 32 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: removed blank space

File: app/src/main/res/layout/state_fragment.xml (hunk 34 above)
Comment (veena14cs) [reply app/src/main/res/layout/state_fragment.xml]: corrected indentation

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt (hunk 35 above)
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt]: Added it for testing purpose. reverted it.

File: app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt (hunk 36 above)
Comment (veena14cs) [reply app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt]: its been removed

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 37 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: removed

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 38 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: rename these variables

**Comment by veena14cs** [comment]
Duplicate of this PR is been created here #[196](https://github.com/oppia/oppia-android/pull/196)
